File: src/main/scala/info/crlog/higgs/protocol/HiggsDecoder.scala
package info.crlog.higgs.protocol

import boson.BosonMessage
import io.netty.handler.codec.frame.FrameDecoder
import io.netty.channel.{Channel, ChannelHandlerContext}
import io.netty.buffer.ChannelBuffer

/**
 * Courtney Robinson <courtney@crlog.rubbish>
 */

class HiggsDecoder extends FrameDecoder {
  override def decode(ctx: ChannelHandlerContext, channel: Channel, buffer: ChannelBuffer): AnyRef = {
    // Wait until the protocol version and message size is available
    if (buffer.readableBytes < 6) {
      return null
    }
    val protocol = buffer.readShort() //protocol version, first 2 bytes
    val messageSize = buffer.readInt() //message size, 3rd to 6th bytes, i.e. 32 bits
    //wait until we get the entire message
    if (buffer.readableBytes < messageSize) {
      buffer.resetReaderIndex()
      return null
    }
    if (protocol == Version.V1.version) {
      buffer.resetReaderIndex()
      return new BosonMessage(buffer.readBytes(messageSize))
    } else {
      throw new UnsupportedVersionException("Version, " + protocol + " is not currently supported")
    }
  }
}

File: src/main/java/info/crlog/higgs/messaging/HiggsDecoder.java
package info.crlog.higgs.messaging;

import io.netty.buffer.ChannelBuffer;
import io.netty.channel.Channel;
import io.netty.channel.ChannelHandlerContext;
import io.netty.handler.codec.oneone.OneToOneDecoder;

/**
 *
 * @author Courtney Robinson <courtney@crlog.info>
 */
public class HiggsDecoder extends OneToOneDecoder {

    protected MessageFactory factory;

    public HiggsDecoder(MessageFactory factory) {
        this.factory = factory;
    }

    @Override
    protected Object decode(ChannelHandlerContext ctx, Channel channel, Object channelBufferMessage) throws Exception {
        if (!(channelBufferMessage instanceof ChannelBuffer)) {
            return channelBufferMessage;
        }
        ChannelBuffer buffer = ((ChannelBuffer) channelBufferMessage);
        ReUsableMessage msg = factory.newMessage();
        int size = msg.getMessageSize(buffer);
        if (size == -1) {
            buffer.resetReaderIndex();
            return null;
        }
        if (buffer.readableBytes() < size) {
            buffer.resetReaderIndex();
            return null;
        }
        buffer.resetReaderIndex();
        //if we get here we've read the entire message
        //the message can populate its internal fields and be returned
        msg.deserialize(buffer);
        return msg;
    }
}
